community/kernel/src/main/java/org/neo4j/kernel/impl/store/countStore/InMemoryCountsStore.java

```diff
@@ -0,0 +1,170 @@
+    private final ConcurrentHashMap<CountsKey,long[]> map;
+    private final OutOfOrderSequence lastTxId = new ArrayQueueOutOfOrderSequence( 0L, 100, EMPTY_METADATA );
+    private CountsSnapshot snapshot;
+
+    public InMemoryCountsStore( CountsSnapshot snapshot )
+    {
+        map = new ConcurrentHashMap<>( snapshot.getMap() );
+        lastTxId.set( snapshot.getTxId(), EMPTY_METADATA );
+    }
+
+    public InMemoryCountsStore()
+    {
+        map = new ConcurrentHashMap<>();
+        lastTxId.set( 0, EMPTY_METADATA );
+    }
+
+    @Override
+    public long[] get( CountsKey key )
+    {
+        return map.get( key );
+    }
+
+    @Override
+    public void updateAll( long txId, Map<CountsKey,long[]> pairs )
+    {
+        lock.readLock().lock();
+        try
+        {
+            applyUpdates( pairs, map );
+            if ( snapshot != null && snapshot.getTxId() >= txId )
+            {
+                applyUpdates( pairs, snapshot.getMap() );
+            }
+            lastTxId.offer( txId, EMPTY_METADATA );
+        }
+        finally
+        {
+            lock.readLock().unlock();
+        }
+    }
+
+    /**
+     *  For each key in the updates param, applies it's value as a diff to the cooresponding value in the given map.
+     * @param updates A map containing the diffs to apply to the corresponding values in the map.
+     * @param map The map to be updated.
+     */
+    private void applyUpdates( Map<CountsKey,long[]> updates, Map<CountsKey,long[]> map )
+    {
+        updates.forEach( ( key, value ) -> map.compute( key, ( k, v ) -> {
+            if ( v == null )
+            {
+                return Arrays.copyOf( value, value.length );
+            }
+            else
+            {
+                return updateEachValue( v, value );
+            }
+        } ) );
+    }
+
+    /**
+     * Because IndexSampleKey and IndexStatisticsKey have 2 values and NodeKey and RelationshipKey have only 1 value,
+     * these keys are stored in the same map by making the value an array of longs. Therefore, when applying the diff
+     * to the keys values, this loop works for keys of both types. In the future, we should separate the values in
+     * IndexSample and IndexStatistics into multiple keys each with one value.
+     */
+    private long[] updateEachValue( long[] v, long[] value )
+    {
+        for ( int i = 0; i < v.length; i++ )
+        {
+            v[i] = v[i] + value[i];
+        }
+        return v;
+    }
+
+    /**
+     * This method is thread safe w.r.t updates to the countstore, but not for performing concurrent snapshots.
+     */
+    @Override
+    public CountsSnapshot snapshot( long txId )
+    {
+        lock.writeLock().lock();
+
+        if ( snapshot != null )
+        {
+            throw new IllegalStateException( "Cannot perform snapshot while another snapshot is processing." );
+        }
```

Comment:
This has to be move in the `try` block of the `try-finally` otherwise on failure you'll never release the lock.